src/core/bpf-firewall.c

```diff
@@ -310,7 +310,7 @@ static int bpf_firewall_compile_bpf(
                         BPF_LD_MAP_FD(BPF_REG_1, accounting_map_fd),
                         BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_map_lookup_elem),
                         BPF_JMP_IMM(BPF_JEQ, BPF_REG_0, 0, 2),
-                        BPF_LDX_MEM(BPF_W, BPF_REG_1, BPF_REG_6, offsetof(struct __sk_buff, len)), /* r1 = skb->len */
+                        BPF_LDX_MEM(BPF_W, BPF_REG_1, BPF_REG_6, offsetof(struct __sk_buff, cb[0])), /* r1 = qdisc_skb_cb(skb)->pkt_len */
                         BPF_RAW_INSN(BPF_STX | BPF_XADD | BPF_DW, BPF_REG_0, BPF_REG_1, 0, 0), /* xadd r0 += r1 */
```

Comment:
i am really irritated by this, no sure how this is supposed to work in the kernel. so kernel headers define cb to be `char[48]`, whiel libbpf headers define it to `u32[5]`, which seems quite a decrapancy to me...

And can we even rely that cb[] is always qdisc private data here?

we really should look into rewriting the program into C, and load it via libbpf, like we are doing in the newer bpf programs...

Reply:
I don't know but probably not.

The netfilter counters undercount in exactly the same way, it's really bothersome.

I think maybe it needs to be fixed in the kernel, and possibly a bpf helper could be added to help here?